core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java

```diff
@@ -0,0 +1,349 @@
+    final boolean executeOnReplicas;
+    final boolean checkWriteConsistency;
+    final Primary<Request, ReplicaRequest, Response> primary;
+    final Replicas<ReplicaRequest> replicasProxy;
+    final AtomicBoolean finished = new AtomicBoolean();
+    final ActionListener<Response> finalResponseListener;
+
+    volatile Response finalResponse = null;
+
+    private final List<ReplicationResponse.ShardInfo.Failure> shardReplicaFailures = Collections.synchronizedList(new ArrayList<>());
+
+
+    ReplicationOperation(Request request, Primary<Request, ReplicaRequest, Response> primary,
+                         ActionListener<Response> listener,
+                         boolean executeOnReplicas, boolean checkWriteConsistency,
+                         Replicas<ReplicaRequest> replicas,
+                         Supplier<ClusterState> clusterStateSupplier, ESLogger logger, String opType) {
+        this.checkWriteConsistency = checkWriteConsistency;
+        this.executeOnReplicas = executeOnReplicas;
+        this.replicasProxy = replicas;
+        this.primary = primary;
+        this.finalResponseListener = listener;
+        this.logger = logger;
+        this.request = request;
+        this.clusterStateSupplier = clusterStateSupplier;
+        this.opType = opType;
+    }
+
+    void execute() throws Exception {
+        final String writeConsistencyFailure = checkWriteConsistency ? checkWriteConsistency() : null;
+        final ShardId shardId = primary.routingEntry().shardId();
+        if (writeConsistencyFailure != null) {
+            finishAsFailed(new UnavailableShardsException(shardId,
+                "{} Timeout: [{}], request: [{}]", writeConsistencyFailure, request.timeout(), request));
+            return;
+        }
+
+        totalShards.incrementAndGet();
+        pendingShards.incrementAndGet(); // increase by 1 until we finish all primary coordination
+        Tuple<Response, ReplicaRequest> primaryResponse = primary.perform(request);
+        successfulShards.incrementAndGet(); // mark primary as successful
+        finalResponse = primaryResponse.v1();
+        ReplicaRequest replicaRequest = primaryResponse.v2();
+        assert replicaRequest.primaryTerm() > 0 : "replicaRequest doesn't have it's primary term" ;
+        if (logger.isTraceEnabled()) {
+            logger.trace("[{}] op [{}] completed on primary for request [{}]", shardId, opType, request);
+        }
+        // we have to get a new state after successfully indexing into the primary in order to honour recovery semantics.
+        // we have to make sure that every operation indexed into the primary after recovery start will also be replicated
+        // to the recovery target. If we use an old cluster state, we may miss a relocation that has started since then.
+        // If the index gets deleted after primary operation, we skip replication
+        List<ShardRouting> shards = getShards(shardId, clusterStateSupplier.get());
```

Comment:
Just wondering if in the future we could use IndexShard as a point to synchronize recovery and replication. An IndexShard representing a primary could also precalculate the shards to replicate to. The code here would then just have to ask IndexShard where to replicate to.